### DISCUSSION TITLE
Migrating from Test Netbox to Live Netbox

### DISCUSSION THREAD

**Discussion by aquelf** [netbox-community/netbox#7494] [Q&A]
I'm sure that this question has been asked many times before but I was just wondering whether I'm on the right path.

We currently run Netbox in a test environment & had planned to lift & shift but it turns out that isn't possible. 

Is it possible to install the Netbox in the new location from start to finish (as oer install guide)

Exporting & Importing Netbox db:

Exporting the db from the current instance using:
pg_dump --username netbox --password --host localhost netbox > netbox.sql

Then loading the Exported db with the following commands?
psql -c 'drop database netbox'
psql -c 'create database netbox'
psql netbox < netbox.sql

Thanks in advance.

**Comment by bitcollector1** [comment]
Looks good.  My command is a little different but I've used it for a few migrations and it worked great each time. 
`pg_dump -h localhost -U netbox -Fc netbox > netbox_db.dump`

**Reply by aquelf** [reply]
Thanks very much for the reply. I'll give it a go & see how I get on, the only thing that is puzzling me is, if I export the existing database from Netbox into a file, where does the new instance know where to pull this file from or do I need to store it somewhere beforehand.

**Reply by bogdancordos** [reply]
Best way will be to copy the _netbox.sql_ to the new instance and run the _psql netbox < netbox.sql_ from its location.

**Comment by aquelf** [comment]
Hi all,

I'm having a few problems trying to get my db imported.

I'm trying this on a test environment, I have set up 2 instances running Netbox. v1 has some data, v2 has nothing. 

When I do a dump on the db using the pg_dump --username netbox --password --host localhost netbox > netbox.sql it copies to file.

I have extracted the file from the cli using WinScp, I then imported it into the new instance of Netbox, however when I try to run the below commands in the cli I get an error saying the user doesn't exist.

psql -c 'drop database netbox'
psql -c 'create database netbox'
psql netbox < netbox.sql

When I try to run the commands in the psql, the commands do not throw up any errors but also do not seem to work. It's most likely something simple, anyone any ideas? 

Thanks in advance.

**Reply by bogdancordos** [reply]
You might need to import the DB as root:

`sudo -u postgres psql netbox < netbox.sql`

**Reply by aquelf** [reply]
Hi Bogdan, that seemed to work a treat. I was trying to do it separately as root & also trying to do it in the postgres. 

Once I rebooted the vm, it cleaned up all old sessions & the db imported correctly without any errors. 

Thank you so much for the assistance.

**Comment by berkeleycowboy** [comment]
I'm planning a migration from Netbox v2.8.8 (PSQL  12.5) a new server running v3.5.8 (PSQL 15.2)

Will I be able to use the pg_dump to migrate the database to a new server ?  

pg_dump -h localhost -U netbox -Fc netbox > netbox_db.dump

**Reply by candlerb** [reply]
I don't think the Postgres versions will be an issue, but you're missing a critical step: you must go via Netbox v2.11.x

Your process will be:
1. Install Netbox v2.11.12 somewhere
2. Dump the original database, import it into the v2.11.12 server, run `./upgrade.sh`, test all is working OK
3. Now dump the database again (from this new server)
4. Install Netbox v3.6.1 somewhere and import the database (or upgrade this one in-place)
5. Run `./upgrade.sh` on here

The only issue I know about for new versions of Postgres is that you might need to do `ALTER DATABASE netbox OWNER TO netbox;` on the new server

EDIT 1: with Netbox v2.11, you may find issues using Python later than 3.8; but this is also the oldest which is compatible with Netbox v3.6. This may affect your choice of VM for Netbox v2.11, or your upgrade plan sequence.

EDIT 2: I suggest you read the vX.Y.0 release notes for *all* the intervening versions, i.e. 2.9.0, 2.10.0, ... 3.6.0, at very least the "Breaking Changes" section. For example, the [v3.0.0](https://github.com/netbox-community/netbox/releases/v3.0.0) release notes are where you would have learned that you can only upgrade from Netbox v2.11. Also you'll learn about cases where data may be dropped.
